dandiapi/api/copy.py

```diff
@@ -0,0 +1,130 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+import math
+
+import boto3
+
+try:
+    from storages.backends.s3boto3 import S3Boto3Storage
+except ImportError:
+    # This should only be used for type interrogation, never instantiation
+    S3Boto3Storage = type('FakeS3Boto3Storage', (), {})
+try:
+    from minio_storage.storage import MinioStorage
+except ImportError:
+    # This should only be used for type interrogation, never instantiation
+    MinioStorage = type('FakeMinioStorage', (), {})
+
+from dandischema.digests.dandietag import PartGenerator, gb, tb
+
+
+@dataclass
+class CopyObjectResponse:
+    key: str
+    etag: str
+
+
+@dataclass
+class CopyPartGenerator(PartGenerator):
+    """Override the normal PartGenerator class to account for larger copy part sizes."""
+
+    @classmethod
+    def for_file_size(cls, file_size: int) -> CopyPartGenerator:
+        """Calculate sequential part sizes given a file size."""
+        if file_size == 0:
+            return cls(0, 0, 0)
+
+        part_size = gb(5)
+
+        if file_size > tb(5):
+            raise ValueError('File is larger than the S3 maximum object size.')
+
+        if math.ceil(file_size / part_size) >= cls.MAX_PARTS:
+            part_size = math.ceil(file_size / cls.MAX_PARTS)
+
+        assert cls.MIN_PART_SIZE <= part_size <= cls.MAX_PART_SIZE
+
+        part_qty, final_part_size = divmod(file_size, part_size)
+        if final_part_size == 0:
+            final_part_size = part_size
+        else:
+            part_qty += 1
+        if part_qty == 1:
+            part_size = final_part_size
+        return cls(part_qty, part_size, final_part_size)
+
+
+def copy_object(
+    storage, source_bucket: str, source_key: str, dest_bucket: str, dest_key: str
+) -> CopyObjectResponse:
+    """Copy an object, returning the new object key and etag."""
+    if isinstance(storage, S3Boto3Storage):
+        client = storage.connection.meta.client
+    elif isinstance(storage, MinioStorage):
+        # Instantiate an s3 client against the minio storage
+        client = boto3.client(
+            's3',
+            endpoint_url=storage.client._endpoint_url,
+            aws_access_key_id=storage.client._access_key,
+            aws_secret_access_key=storage.client._secret_key,
+            region_name='us-east-1',
+        )
+    else:
+        raise ValueError(f'Unknown storage {storage}')
+
+    return copy_object_multipart(
+        client,
+        source_bucket=source_bucket,
+        source_key=source_key,
+        dest_bucket=dest_bucket,
+        dest_key=dest_key,
+    )
+
+
+def copy_object_multipart(
```

Comment:
This is technically also valid for minio, which is why I named is as such. I suppose if we consider s3 against minio as s3 all the same, then your change makes sense.

Reply:
I thought it make sense to label it S3 since it uses a boto3 client. If we had a `_copy_object_multipart_minio` which used a Minio client, we could obviously still use it with an S3 bucket, but I think it still makes sense to clarify with flavor of client is being used. Just my 2¢